pkg/template/apis/template/constants.go

```diff
@@ -39,4 +39,12 @@ const (
 	// BindableAnnotation indicates whether the template service broker should
 	// advertise the template as being bindable (default is true)
 	BindableAnnotation = "template.openshift.io/bindable"
+
+	// TemplateInstanceFinalizer is used to clean up the objects created by the template instance,
+	// when the template instance is deleted.
+	TemplateInstanceFinalizer = "template.openshift.io/finalizer"
+
+	// TemplateInstanceOwner is a label applied to all objects created from a template instance
+	// which contains the uid of the template instance.
+	TemplateInstanceOwner = "template.openshift.io/template-instance-owner"
```

Comment:
again reading top to bottom, but this screams: "migration problem" to me.

Reply:
templateinstances created by old versions won't have this finalizer, and resources created by those templateinstances will have ownerrefs on them, which means they'll get handled by normal GC and the finalizer should ignore them.

templateinstances created by this version will have the finalizer and the objects created by the templateinstance will have the label and no ownerref, so GC will ignore them and finalization will cover it.

At least that is my theory.  Other than "i was creating a template instance but it's objects weren't created yet and then i upgraded my controller" do you see a hole?  I don't think i can convince myself to care about that scenario.

Reply:
I thought this was solving a cross namespace problem for you.  Don't you need to have the cluster update to be in a state where it's solved?

Reply:
not really, no.  anyone who created cross-namespace content already (which i suspect is no one) has already potentially experienced that content getting spontaneously erased by the GC (since as far as the GC is concerned, that content has no valid ownerref already).

We might do a knowledge base about cleaning up that potential content, but again i think cross-namespace object creation is a really rare use case.  In particular because it would have to be coupled with either TSB based instantiation, or direct templateinstance creation, which no one knows how to do except us.